xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,271 @@
+  private final LoadBalancerRegistry lbRegistry;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
+
+  // The following fields become non-null once handleResolvedAddresses() successfully.
+
+  // Most recent CdsConfig.
+  @Nullable
+  private CdsConfig cdsConfig;
+  // Most recent ClusterWatcher.
+  @Nullable
+  private ClusterWatcher clusterWatcher;
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  @Nullable
+  private XdsClient xdsClient;
+
+  CdsLoadBalancer(Helper helper) {
+    this(helper, LoadBalancerRegistry.getDefaultRegistry());
+  }
+
+  @VisibleForTesting
+  CdsLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry) {
+    this.channelLogger = helper.getChannelLogger();
+    this.lbRegistry = lbRegistry;
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(helper);
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+    Attributes attributes = resolvedAddresses.getAttributes();
+    if (xdsClientRef == null) {
+      xdsClientRef = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_REF);
+      if (xdsClientRef == null) {
+        // TODO(zdapeng): create a new xdsClient from bootstrap if no one exists.
+        throw new UnsupportedOperationException(
+            "XDS_CLIENT_REF attributes not available in resolvedAddresses " + resolvedAddresses);
+      }
+      xdsClient = xdsClientRef.getObject();
+    }
+
+    Map<String, ?> newRawLbConfig = checkNotNull(
+        attributes.get(ATTR_LOAD_BALANCING_CONFIG), "ATTR_LOAD_BALANCING_CONFIG not available");
+    ConfigOrError cfg =
+        CdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(newRawLbConfig);
+    if (cfg.getError() != null) {
+      throw cfg.getError().asRuntimeException();
+    }
+    final CdsConfig newCdsConfig = (CdsConfig) cfg.getConfig();
+
+    // If CdsConfig is changed, do a graceful switch.
+    if (!newCdsConfig.equals(cdsConfig)) {
+      LoadBalancerProvider fixedCdsConfigBalancerProvider =
+          new FixedCdsConfigBalancerProvider(newCdsConfig);
+      switchingLoadBalancer.switchTo(fixedCdsConfigBalancerProvider);
+    }
+
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+
+    // The clusterWatcher is also updated after switchingLoadBalancer.handleResolvedAddresses().
+    cdsConfig = newCdsConfig;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.ERROR, "Name resolution error: '%s'", error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
```

Comment:
Just to be pedantic: do you also want to check that `xdsClient != null`? And inside the `if` block set `xdsClient = null`  ?

Reply:
The `returnObject()` is required to be called with the right `xdsClient` argument, if the argument passed in is null or not the one returned from `xdsClientRef.get()`, it will throw. It would be a bug if `xdsClient == null` at this point.